lxd/instance/drivers/driver_qemu.go

```diff
@@ -7511,6 +7533,8 @@ func (d *qemu) nextVsockID() (uint32, error) {
 		}
 
 		if d.freeVsockID(candidateVsockID) {
+			nextVsockIDMutex.Lock()
+			go waitForVsockID(d.state.ShutdownCtx, candidateVsockID)
```

Comment:
I dont see what the need for the go routine is here, it seems overly complicated.
Adding the check for the context being cancelled in the loop makes sense though.
Also I think the whole qemu.nextVsockID needs to be protected by the mutex, otherwise the first call to 	d.freeVsockID(vsockID) is racy also.

Reply:
If we just lock down `nextVsockID` we unlock before the VM was started. This means even the vsock ID isn't bound yet so the next caller to `nextVsockID` will see that the said vsock ID is still free and get it as well. The goroutine should unlock until the vsock ID is actually acquired.

Reply:
Mh yes. At least `d.freeVsockID(candidateVsockID)` needs to be put behind the lock, otherwise the next caller will already have the same vsock ID, but waits for the lock to come back to see that it already is acquired.

Reply:
Hrm, yeah thats going to bound the concurrent start times of VMs which wouldn't be good.
Maybe we can keep track of vsock ids already allocated to starting VMs in a slice (whos access is also controlled by that mutex) and skip those, and then when the vm is started (or fails to start) we clear that entry from the list?

Either that or we could explore opening the vsock and then passing the fd to qemu to use, so it remains open from that point onwards.

Reply:
Actually, we will always face an issue if we "pre allocate" even if we store it in a slice. There is some time in between allocation and when the VM actually starts listening on the vsock ID. Especially in nested LXD scenarios, this might become a problem.

I'll investigate passing the fd.

Reply:
@tomponline This went quite good, thanks for the hint from Bugzilla, looks promising, please have another look.

Also I have the feeling that this vsock selection is now much faster since we don't have to call the library function anymore that tries to open up a connection to this socket. It's solved by a single syscall.